### PR TITLE
✨ Cancun ReentrancyGuard

### PR DESCRIPTION
**Pull request by Vectorized**
## Description

Waiting for: https://github.com/foundry-rs/compilers/issues/46

## Checklist

Ensure you completed **all of the steps** below before submitting your pull request:

- [x] Ran `forge fmt`?
- [x] Ran `forge snapshot`?
- [x] Ran `forge test`?

_Pull requests with an incomplete checklist will be thrown out._

### CODE REVIEW

File: src/utils/ReentrancyGuard.sol [hunk 1]
```diff
@@ -31,24 +31,24 @@ abstract contract ReentrancyGuard {
...
                 revert(0x1c, 0x04)
             }
-            sstore(_REENTRANCY_GUARD_SLOT, 2)
+            tstore(_REENTRANCY_GUARD_SLOT, 2)
         }
         _;
         /// @solidity memory-safe-assembly
         assembly {
-            sstore(_REENTRANCY_GUARD_SLOT, 1)
+            tstore(_REENTRANCY_GUARD_SLOT, 1)
```
Comment (hrkrshnn) [inline src/utils/ReentrancyGuard.sol]: Why not 0 and 1? Should be cheaper due to push0 and the odd storage pricing shouldn't matter here IIRC.

Comment (Vectorized) [reply src/utils/ReentrancyGuard.sol]: Thx!

File: src/utils/ReentrancyGuard.sol [hunk 2]
```diff
@@ -31,24 +31,24 @@ abstract contract ReentrancyGuard {
     modifier nonReentrant() virtual {
         /// @solidity memory-safe-assembly
         assembly {
-            if eq(sload(_REENTRANCY_GUARD_SLOT), 2) {
+            if tload(_REENTRANCY_GUARD_SLOT) {
                 mstore(0x00, 0xab143c06) // `Reentrancy()`.
                 revert(0x1c, 0x04)
             }
-            sstore(_REENTRANCY_GUARD_SLOT, 2)
+            tstore(_REENTRANCY_GUARD_SLOT, address())
```
Comment (MathisGD) [inline src/utils/ReentrancyGuard.sol:38]: curiosity: why pushing the address and not 1 here?

Comment (Vectorized) [reply src/utils/ReentrancyGuard.sol:38]: Save gas.

`ADDRESS` (2 gas, 1 byte). `PUSH1 0x01` (3 gas, 2 bytes).

**Pull request merged**